### ISSUE TITLE
[Scala] Consolidate `scala`, `async-scala` into a single Scala API client

### ISSUE THREAD

**Issue by wing328** [swagger-api/swagger-codegen#3812]
##### Description

Similar to Java, C#, Python API clients, we want to be able to provide a single Scala API client to perform synchronous and asynchronous HTTP calls.

##### Swagger-codegen version

Latest master

If anyone from the community wants to work on the consolidation, please refer to let us know.

A good starting point is to review how Java API client supports both synchronous and asynchronous calls and implement the same logic in the `scala` or `async-scala` generators and templates.

**Comment by wing328** [comment]
UPDATE: `scala` API client in 2.3.0 now supports async request.

The next step is to update `scala` to use akka-http to make HTTP calls.

**Comment by wing328** [comment]
A test has been commented out due to issue mentioned in https://github.com/swagger-api/swagger-codegen/pull/5115#issuecomment-290912839

Updating HTTP lib to use `akka-http` will fix it.

**Comment by foxmk** [comment]
Hello! Is there any WIP on this issue? If not, I will take it

**Comment by wing328** [comment]
@foxmk thanks for offering help on this. In the 2.3.0 branch, `async-scala` and `scala` generator has been merged into one thanks to the good work from @clasnake 

If we can update `scala` generator in 2.3.0 to generate Scala code with the `akka-http` lib, that should fix the issue mentioned in https://github.com/swagger-api/swagger-codegen/pull/5115#issuecomment-290912839

**Comment by wing328** [comment]
Here are some good starting points:

ApiInvoker mustache template for `scala` (2.3.0): https://github.com/swagger-api/swagger-codegen/blob/2.3.0/modules/swagger-codegen/src/main/resources/scala/apiInvoker.mustache

ApiInvoker mustache template for `akka-scala` (2.3.0): https://github.com/swagger-api/swagger-codegen/blob/2.3.0/modules/swagger-codegen/src/main/resources/akka-scala/apiInvoker.mustache

**Comment by foxmk** [comment]
Great! I have bunch of questions:
- If I understood correctly, we should have only one Scala client generator?
- Which project should we take as base, as they are quite different? `akka-scala` is Maven-only project with `spray-json` (now `akka-http`) and `json4s`. `scala` generator uses Jackson and Jersey as HTTP client. I assume It’s better to take `akka-scala` as base (and probably call it `scala`).
- Should we use `spray-json` for serialization, as it is kinda standart for `akka-http` projects (and faster)?
- Are there any plans to make `akka-http-server` generator? I can take it first to get familiar with codebase and then come back to this issue.

**Comment by wing328** [comment]
> If I understood correctly, we should have only one Scala client generator?

Yes, that's the goal as one generator will be easier to maintain than having 3 generators doing essentially the same thing.

> Which project should we take as base

For the `akka-scala` generator, it's not very mature compared with the `scala` generator. in the [Petstore sample](https://github.com/swagger-api/swagger-codegen/tree/2.3.0/samples/client/petstore/akka-scala/src/main), there's no integration tests so I'm not sure if the basic use cases work. For Scala Petstore sample (2.3.0), you can find the integration tests in https://github.com/swagger-api/swagger-codegen/tree/2.3.0/samples/client/petstore/scala/src/test/scala.

That's why we've consolidated `async-scala` into `scala` 

> Should we use spray-json for serialization,

Sounds good to me.

> Are there any plans to make akka-http-server generator? 

I don't think so. I've opened a [ticket](https://github.com/swagger-api/swagger-codegen/issues/5900) for tracking and see if anyone has started working on something similar or want to help out.

A good starting point is to take a look at other server generator PRs to see what files need to be added/modified to create the new generator: https://github.com/swagger-api/swagger-codegen/pulls?q=is%3Apr+is%3Aclosed+label%3A%22Enhancement%3A+New+generator%22

**Comment by clasnake** [comment]
@foxmk Thanks very much. In terms of serialization library, spray-json is very convenient to use for normal types, but it's quite an opinionated library so I don't mind if you choose other libraries such as jackson if that's easy to deal with from the mustache template design point of view.

**Comment by foxmk** [comment]
@clasnake Yes, I would agree that `spay-json` can be tricky in codegen.

I've started digging down into code generation process and I have more questions:
- There is still `async-scala` generator and example code for it. Should I just ignore it? Or remove?
- As I understood, tests for API examples are written manually, am I right? I found no templates for them.
- The whole chain of call from API class to `ApiInvoker` to `com.wordnik.swagger.client.ApiClient` etc. seems a bit overcomplicated. Is it intentional, or I miss something? (Probably this question is too broad, so I wouldn't mind if you just say that it is required and I shouldn't remove it)
- Also, do we have Gitter chat or smth like that?

**Comment by wing328** [comment]
> There is still async-scala generator and example code for it. Should I just ignore it? Or remove?

I'll remove it later from 2.3.0

> As I understood, tests for API examples are written manually, am I right? I found no templates for them.

Yes, for other langauges like C#, Ruby, Java, we've added auto-generated test cases and we can implement it later for Scala as a day 2 requirement.

> The whole chain of call from API class to ApiInvoker to com.wordnik.swagger.client.ApiClient

That's the current design. Using Java as an example, here is how the request works:

addPet (in PetApi.java) => ApiClient.java => HTTP libraries (e.g. Jersey, Jersey2.x, Retrofit, Retrofit2, Feign, etc) => HTTP request

The ApiClient.java is "swappable" allowing as to easily support more than 5 different Java HTTP. libraries via the `library` option:
```
	library
	    library template (sub-template) to use (Default: okhttp-gson)
	        jersey1 - HTTP client: Jersey client 1.19.1. JSON processing: Jackson 2.7.0. Enable Java6 support using '-DsupportJava6=true'. Enable gzip request encoding using '-DuseGzipFeature=true'.
	        feign - HTTP client: OpenFeign 9.4.0. JSON processing: Jackson 2.8.7
	        jersey2 - HTTP client: Jersey client 2.22.2. JSON processing: Jackson 2.7.0
	        okhttp-gson - HTTP client: OkHttp 2.7.5. JSON processing: Gson 2.6.2. Enable Parcelable modles on Android using '-DparcelableModel=true'. Enable gzip request encoding using '-DuseGzipFeature=true'.
	        retrofit - HTTP client: OkHttp 2.7.5. JSON processing: Gson 2.3.1 (Retrofit 1.9.0). IMPORTANT NOTE: retrofit1.x is no longer actively maintained so please upgrade to 'retrofit2' instead.
	        retrofit2 - HTTP client: OkHttp 3.2.0. JSON processing: Gson 2.6.1 (Retrofit 2.0.2). Enable the RxJava adapter using '-DuseRxJava[2]=true'. (RxJava 1.x or 2.x)
	        resttemplate - HTTP client: Spring RestTemplate 4.3.7-RELEASE. JSON processing: Jackson 2.8.8
	        resteasy - HTTP client: Resteasy client 3.0.19.Final. JSON processing: Jackson 2.7.0
```
> Also, do we have Gitter chat or smth like that?

For Swagger Codegen, we primarily rely on Swagger Codegen issue tracker for discussion.

(you can reach me via google chat through my email address listed in my Github profile page if you want to have a quick chat over IM)

**Comment by wing328** [comment]
@foxmk do you have any more questions? If you need any help, please let us know.

**Comment by foxmk** [comment]
@wing328 Hello! Sorry for being slowpoke. So far I have no questions, I have a rough plan to solve this issue, similar to Java generator:
- Abstract current Jetty client behind new API with its own HTTP model (in progress)
- Make akka-http implementation of HTTP client alongside with old Jetty

But it's hard with timing, I think I will finish PR till the end of next week. If it's too long, I can clean up push what I have handle this issue to somebody else.

**Comment by wing328** [comment]
@foxmk thanks for the update. Please take your time and let us know if you need help (e.g. testing). Appreciate your effort on this 👍

**Comment by emtiazahmed** [comment]
How do I use tell maven-plugin to generate async scala client?
I am using the following configuration:
`<plugin>
                <groupId>io.swagger</groupId>
                <artifactId>swagger-codegen-maven-plugin</artifactId>
                <version>2.3.0-SNAPSHOT</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <inputSpec>myspec.yaml</inputSpec>
                            <language>scala</language>
                             <templateDirectory>src/main/resources/templates</templateDirectory>
                            <configOptions>
                                <sourceFolder>src/main/scala</sourceFolder>
                            </configOptions>
                        </configuration>
                    </execution>
                </executions>
            </plugin>`

**Comment by wing328** [comment]
I would suggest a plan B: 
- rename `scala` to `scala-wordnik` in case Scala developers still want to use it (but we will remind them this generator is no longer actively maintained)
- rename `akka-scala` to just `scala` and suggest the community to work on this generator moving forward

cc @ramzimaalej

**Comment by ramzimaalej** [comment]
@foxmk are you actively working on this? If yes, do you need help?

I think we should follow what you did to the Java generator, have only one generator with the option to select desired libraries.
